### PR TITLE
readme.md: clarify behavior of the plugin

### PR DESCRIPTION
**Pull request by waldyrious**
### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 
*   [x] If applicable, I’ve added docs and tests

### Description of changes

It wasn't immediately obvious how this plugin worked. I've added a bit more detail in the README to expose that information more clearly.

### CODE REVIEW

**Review by Murderlon — COMMENTED** [review]
Hi, I'm not sure if this is an improvement, especially this sentence feels a bit off

> It requires the section heading to be present in the source

I'd say we keep the "What is this" section for explaining what this package does conceptually, and leave the instructions for "Install" and "Options" (where both of your additions are already explained).

**Comment by waldyrious** [comment]
Thanks for the review. It's not entirely clear to me what you're suggesting, though. How do you propose that I proceed?

**Comment by Murderlon** [comment]
Given the fact that this is also presented in the next section:

> [...] When this plugin is used, authors can add a certain heading (say, `## License`)
to documents and this plugin will populate them.

I'd say it's clear what this plugin does and when to use it. Unless you have something specific in mind that is missing and is needed to understand the plugin, I think I'll close this

**Comment by waldyrious** [comment]
Gotcha, thanks for clarifying. I would say that the current text "generate a license section" is a bit too hand-wavy as an explanation of what this package does conceptually, to use your words. Even the description in the second section doesn't make it clear where the information comes from.

Let me try reworking the PR in light of what you said; I think I might be able to phrase it in a way that surfaces the info upfront without compromising the spirit of those sections as you've described them.

**Review by wooorm — COMMENTED** [review]
I think “What is this?” should remain relatively short.
`remark-toc` is an example with some more explanation: https://github.com/remarkjs/remark-toc/blob/main/readme.md. You could try something like that

**Comment by waldyrious** [comment]
I've updated the PR – please take a look and let me know what you think! :)

File: readme.md [hunk 1]
```diff
@@ -27,8 +27,8 @@
 
 ## What is this?
 
-This package is a [unified][] ([remark][]) plugin to generate a license section
-such as the one below.
+This package is a [unified][] ([remark][]) plugin to fill in a license section
```
Comment (Murderlon) [inline readme.md:30]: I think "generate" was more accurate in this context

File: readme.md [hunk 2]
```diff
@@ -27,8 +27,8 @@
...
+such as [the one](#license) at the end of this README.
```
Comment (Murderlon) [inline readme.md:31]: I can agree to this change

File: readme.md [hunk 3]
```diff
@@ -42,8 +42,11 @@ This project is useful when you’re writing documentation for an open source
 project, typically a Node.js package, that has one or more readmes and maybe
 some other markdown files as well.
 You want to show the author and license associated with the project.
+
 When this plugin is used, authors can add a certain heading (say, `## License`)
-to documents and this plugin will populate them.
+to documents and this plugin will populate that section with the information,
+using either the `author` and `license` fields set in a `package.json` file,
+or parameters with the same name passed to the plugin.
```
Comment (Murderlon) [inline readme.md:49]: I don't think this is much of an improvement. 

- The section is about "when should I use this" and less about implementation details, which are documented below. 
- "with the information" feels a bit off sentence wise and doesn't refer back to something what that information is
- "using either the `author` and `license`" — either is not correct here, it's not one or the other, they are both used for different things.

**Comment by Murderlon** [comment]
I'm closing this as I don't see we are headed to an improvement we agree on. However I do appreciate the effort you put in helping us improve.

**Comment by waldyrious** [comment]
I would be willing to engage with the feedback you provided and make adjustments toward something more agreeable to your vision of what the sections should say, but I guess it's best to wrap this up if you are more inclined to reject this change's overall direction, and spare us both the energy.

I just want to reiterate the original motivation for this PR, which is that it's not stated up-front what exactly this plugin does (the intro sections are, as mentioned above, a bit hand-wavy), and that it would be better to not force people to read from multiple sections of the README just to they figure out whether this behaves in a way that would be useful to them. Please consider revising the README if there's any change in wording that would be acceptable to you while making the at-a-glance information a little more explicit.

**Comment by wooorm** [comment]
Perhaps we can understand you better if you can elaborate on how this plugin behaves differently than what you expected?

**Comment by waldyrious** [comment]
Thanks for asking. The issue is precisely that I didn't know (at first) what to expect. The repository tagline (which is also the first sentence of the README) says:

> plugin to generate a license section 

The "What is this?" section says that the plugin serves:

> ...to generate a license section such as the one below.

The "When should I use this?" section says:

> to show the author and license associated with the project (...) authors can add a certain heading (say, ## License) to documents and this plugin will populate them.

Honestly, the "add a certain heading" part went over my head the first time I skimmed the project, since it was tucked inside a paragraph that seemed to be about why one would want a license section in a README, which I already knew I'd want.

So my doubts were two-fold: how the plugin would know where to inject the licensing section, and how it would know what information to put there. Now, to be fair, both of these details _are_ present in the README, but they are scattered throughout it (the concrete example of the plugin's output, in particular, is at the very bottom!). They are not surfaced in an easily glanceable way, and therefore what someone would grasp right away (when quickly assessing the project) might end up sounding a little like magic.

I just thought it would be helpful to expose a more concrete image of what this does in a way that's accessible to a first skim of the README, rather than expect people to dive deep into the README just to gain some sense of whether this is would be useful to them. It's just a small QOL improvement to the first-time experience of engaging with this project, so it's not a big deal and we've probably expended more energy discussing this PR than what would be reasonable given the benefits it would (IMHO) bring — but as they say, first impressions count! :wink:

**Comment by wooorm** [comment]
> how the plugin would know where to inject the licensing section, and how it would know what information to put there.

These are all logical questions. But we can’t explain the whole plugin and all options in one sentence. Or in one paragraph. By definition, that would make that one sentence/paragraph too complex. That why this readme starts of small and expands in-depth the further you read.

> Now, to be fair, both of these details are present in the README, but they are scattered throughout it

- Where to inject is shown in “Use” and explained in “`unified().use(remarkLicense[, options])`” in “API”.
- What to inject is noted in “Use” and explained in-depth in “`options`” in “API”.

I can understand the term “scattered” to describe things being explained multiple times, but your changes add *more* “scattering”. The byline, “What is this?”, and “When should I use this?” sections are not about how to configure the plugin as a developer. That’s where the Use and API sections are for.

> the concrete example of the plugin's output, in particular, is at the very bottom!

There is a “Use” section. The fact that `## License` is at the bottom of readmes has to do with how they are used in open source projects. `remark-toc` uses a heading that’s typically higher. This order is [somewhat “semistandard”](https://github.com/RichardLitt/standard-readme).

> They are not surfaced in an easily glanceable way, and therefore what someone would grasp right away (when quickly assessing the project) might end up sounding a little like magic.

I do not understand this conclusion. As shown above, they’re mentioned several times. Not at the start though. I’d then assume that folks would think: “I should continue reading the docs” rather than “let’s not read the rest of the readme, it must be magic”.

**Comment by waldyrious** [comment]
Yeah, fair points. Just for clarification, what I meant by "scattered" is closer to "non-linear" than to "appearing in multiple places". But I understand your perspective and it makes sense that information duplication should be avoided.

> > They are not surfaced in an easily glanceable way, and therefore what someone would grasp right away (when quickly assessing the project) might end up sounding a little like magic.
>
> I do not understand this conclusion. As shown above, they’re mentioned several times. Not at the start though. I’d then assume that folks would think: “I should continue reading the docs” rather than “let’s not read the rest of the readme, it must be magic”.

I can imagine that my experience was likely atypical: I was just starting to learn about remark, and tried to look at a few plugins that sounded interesting from the name + tagline; I opened several browser tabs and only dedicated a few brain cycles to each of them before deciding whether to explore further or abandon them. This one happened to offer a mix of information and promise that made it sit in the fence between those two outcomes, so I eventually came back to it to read the docs more in depth. I wished I didn't have to keep that task in the backburner while I was processing other information, so I suggested this PR with the changes I assumed would have helped me get to a decision quicker.

That said, I assumed that making the "quick-glance" information a little more explicit would benefit everyone (not unlike how accessibility improvements end up making everyone's experience better). I know that the readme is structured in a progressive manner already, but there are small tweaks that IMO would help it provide the information more effectively — subtle stuff like using a term like "fill in" instead of "generate", perhaps splitting the "add a certain heading" into a separate paragraph in the "When should I use this?" section, explicitly mention the presence of the "License" header in the example document of the Use section, making the word "below" in the "What is this?" section into a link to the license section's anchor, etc.

It's arguable that such changes would only provide marginal improvements, so just to be clear, I do understand if you think it's preferable to leave the readme as it is than to bikeshed minute improvements to it. The time you both have spent already engaging this discussion is well appreciated regardless :)

**Comment by wooorm** [comment]
Thanks for explaining your thought process!
I had a look again and added a link to the license section, and made sure you’re credited as a co-author!
I’m not, and neither was @Murderlon before, in favor of any of the other proposed changes. I don’t think they’re improvements, and I don’t have alternative ideas on what wording to use which I would see as an improvement.

**Comment by waldyrious** [comment]
Thanks! I appreciate the change and the mention, as well as the time and patience you both expended in this issue. Keep up the great work!

**Pull request closed without merging**